CRM/Civicase/Service/CaseTypeCategoryEventHandler.php

```diff
@@ -0,0 +1,236 @@
+      return;
+    }
+
+    $updateParams = [];
+    if (isset($category['is_active'])) {
+      $updateParams['is_active'] = !empty($category['is_active']) ? 1 : 0;
+    }
+    if (isset($category['icon'])) {
+      $updateParams['icon'] = 'crm-i ' . $category['icon'];
+    }
+
+    if (!empty($updateParams)) {
+      $this->menu->updateItems($category['id'], $updateParams);
+    }
+  }
+
+  /**
+   * Removes case type category menu item from the civicrm navigation bar.
+   *
+   * @param string $name
+   *   Case type category name.
+   */
+  public function onDelete($name) {
+    if (!$name || !$this->shouldRun()) {
+      return;
+    }
+
+    $this->menu->deleteItems($name);
+    $this->customFieldExtends->delete($name);
+    $this->customData->delete($name);
+  }
+
+  /**
+   * Set items to check before running action (onCreate, onUpdate, onDelete).
+   *
+   * You may use this method to override default check settings and provide
+   * necessary data.
+   *
+   * @param array $params
+   *   Parameters that are used to perform necessary checks.
+   */
+  public function setCheck(array $params) {
+    if (isset($params['needCheckReferer'])) {
+      $this->needCheckReferer = $params['needCheckReferer'] ? TRUE : FALSE;
+    }
+    if (isset($params['needCheckOptionGroup'])) {
+      $this->needCheckOptionGroup = $params['needCheckOptionGroup'] ? TRUE : FALSE;
+    }
+    if (isset($params['optionGroupId'])) {
+      $this->optionGroupId = intval($params['optionGroupId']);
+    }
+    if (isset($params['objectName'])) {
+      $this->objectName = $params['objectName'];
+    }
+  }
+
+  /**
+   * Determines if the hook will run.
+   *
+   * Keep in mind that referer will be checked if $needCheckReferer property
+   * is set to TRUE.
+   * Keep in mind that option group will be checked if $needCheckOptionGroup
+   * is TRUE and $optionGroupId is set.
+   * By default this method would always return TRUE.
+   *
+   * @see setCheck()
+   *
+   * @return bool
+   *   TRUE if passed object is an option value of Case Type Categories
+   *   option group and http referer is correct, FALSE otherwise.
+   */
+  protected function shouldRun() {
+    return ($this->needCheckReferer ? $this->checkReferrer() : TRUE)
+      && ($this->needCheckOptionGroup && $this->optionGroupId ? $this->checkOptionGroup($this->optionGroupId, $this->objectName) : TRUE);
+  }
+
+  /**
+   * Checks if OptionValue belongs to Case Type Categories option group.
+   *
+   * @param int|string $optionGroupId
+   *   Case type category option group id or name. If is id then name would be
+   *   fetched from DB.
+   * @param string $objectName
+   *   (Optional) Object name, may be used to check object if it's type is
+   *   unpredictable or unknown.
+   *
+   * @return bool
+   *   TRUE if object is an option value of Case Type Categories option group,
+   *   FALSE otherwise.
+   */
+  protected function checkOptionGroup($optionGroupId, $objectName = 'OptionValue') {
```

Comment:
We don't need to check this. This check seem to be related to the postprocess hook class. It is the responsibility of that class(I think it already does that) to make the check that the option group is for the case type category class. 
This class is basically just a service that performs post process actions for case categories, we should not bloat it with checks that should be made in the class calling the service.

Reply:
Actually this would be required in the listener class as it would be run for OptionValues of all types.

Reply:
This should only be in the listener class then and not in the Event handler class.